e2e-tests/desktop/tests/pagination.spec.js

```diff
@@ -0,0 +1,151 @@
+    hostIds: hosts.map((host) => host.id),
+  });
+
+  // Create target
+  targetWithManyHosts = await boundaryHttp.createTarget(request, {
+    scopeId: project.id,
+    type: 'tcp',
+    port: targetPort,
+  });
+  targetWithManyHosts = await boundaryHttp.addHostSource(request, {
+    target: targetWithManyHosts,
+    hostSourceIds: [hostSet.id],
+  });
+});
+
+test.afterEach(async ({ request }) => {
+  if (org) {
+    await boundaryHttp.deleteOrg(request, org.id);
+  }
+});
+
+test.describe('Target with many hosts', () => {
+  // TODO: Add tests for searching host sources
+  test('should display all hosts in the target', async ({ authedPage }) => {
+    await authedPage
+      .getByRole('link', { name: targetWithManyHosts.name })
+      .click();
+
+    // Check pagination
+    await expect(
+      authedPage.getByRole('cell', { name: hosts[0].name, exact: true }),
+    ).toBeVisible();
+    await expect(
+      authedPage.getByRole('cell', {
+        name: hosts[hosts.length - 1].name,
+        exact: true,
+      }),
+    ).toBeHidden();
+
+    // Navigate to the second page. The last host should now be visible.
+    await authedPage
+      .getByRole('navigation', { name: 'pagination' })
+      .getByRole('button', { name: 'page 2' })
+      .click();
+    await expect(
+      authedPage.getByRole('cell', { name: hosts[0].name, exact: true }),
+    ).toBeHidden();
+    await expect(
+      authedPage.getByRole('cell', {
+        name: hosts[hosts.length - 1].name,
+        exact: true,
+      }),
+    ).toBeVisible();
+
+    // Use the "previous page" button to navigate back to the first page.
+    await authedPage
+      .getByRole('navigation', { name: 'pagination' })
+      .getByRole('button', { name: 'Previous page' })
+      .click();
+    await expect(
+      authedPage.getByRole('cell', { name: hosts[0].name, exact: true }),
+    ).toBeVisible();
+    await expect(
+      authedPage.getByRole('cell', {
+        name: hosts[hosts.length - 1].name,
+        exact: true,
+      }),
+    ).toBeHidden();
+
+    // Use the "next page" button to navigate to the second page again.
+    await authedPage
+      .getByRole('navigation', { name: 'pagination' })
+      .getByRole('button', { name: 'Next page' })
+      .click();
+    await expect(
+      authedPage.getByRole('cell', { name: hosts[0].name, exact: true }),
+    ).toBeHidden();
+    await expect(
+      authedPage.getByRole('cell', {
+        name: hosts[hosts.length - 1].name,
+        exact: true,
+      }),
+    ).toBeVisible();
+
+    // Use the "Items per page" options to show 30 items per page.
+    await authedPage
+      .getByRole('combobox', { name: 'Items per page' })
+      .selectOption('30');
+    await expect(
+      authedPage.getByRole('cell', { name: hosts[0].name, exact: true }),
+    ).toBeVisible();
+    await expect(
+      authedPage.getByRole('cell', {
+        name: hosts[hosts.length - 1].name,
+        exact: true,
+      }),
+    ).toBeVisible();
```

Comment:
Couldn't we assert that there's 15 items now instead? And for the 10 items option that there's 10 rows? Rather than just looking at the first and last item

Reply:
Yeah that would be a better approach. I will update that now.